### PR TITLE
Vault integration: revoke lease

### PR DESCRIPTION
**Pull request by mgaffney**

### CODE REVIEW

**Review by louisruch — COMMENTED** [review]
Couple of quick comments, I was going to review this all but `internal/credential/vault/vault_capabilities.go` was a bit more involved than I would like for a Sunday night :) so I will take a look at that and the test tomorrow morning

File: internal/credential/vault/testing_test.go [hunk 1]
```diff
@@ -157,15 +157,18 @@ func TestTestVaultServer_CreateToken(t *testing.T) {
 			assert, require := assert.New(t), require.New(t)
 			v := NewTestVaultServer(t)
 			require.NotNil(v)
-			secret := v.CreateToken(t, tt.opts...)
+			secret, token := v.CreateToken(t, tt.opts...)
 			require.NotNil(secret)
+			require.NotEmpty(token)
```
[the remark this answers is not in the document]
Comment (mgaffney) [reply internal/credential/vault/testing_test.go:162]: It does, but this is the test for the `CreateToken` test helper method so I figured this was testing the test.

File: internal/credential/vault/vault.go [hunk 2]
```diff
@@ -137,6 +141,17 @@ func (c *client) renewLease(leaseId string, leaseDuration time.Duration) (*vault
 	return t, nil
 }
 
+// revokeLease calls the /sys/leases/revoke Vault endpoint. This endpoint
+// is NOT accessible with the default policy in Vault 1.7.0. See
```
[the remark this answers is not in the document]
Comment (mgaffney) [reply internal/credential/vault/vault.go]: Ty!

File: internal/credential/vault/vault.go [hunk 3]
```diff
@@ -181,3 +196,46 @@ func (c *client) post(path string, data []byte) (*vault.Secret, error) {
 	}
 	return s, nil
 }
+
+// capabilities calls the /sys/capabilities-self Vault endpoint and returns
+// the vault.Secret response. This endpoint is accessible with the default
+// policy in Vault 1.7.0. See
+// https://www.vaultproject.io/api-docs/auth/token#renew-a-token-self.
```
[the remark this answers is not in the document]
Comment (mgaffney) [reply internal/credential/vault/vault.go]: Ty!

File: internal/credential/vault/vault_capabilities.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+		caps = append(caps, "sudo")
+	}
+	if c&rootCapability == rootCapability {
+		caps = append(caps, "root")
+	}
+
+	return caps
+}
+
+func (c capabilities) String() string {
```
Comment (louisruch) [inline internal/credential/vault/vault_capabilities.go:114]: This is fine, but doing what you did in `GoString` and using strings.Join seems a bit more readable to me.  So something like:
```
	if c == noCapabilities {
		return "[]"
	}
	var b strings.Builder
	b.WriteString(`["`)
	b.WriteString(strings.Join(caps, `" ,"`))
	b.WriteString(`"]`)
	return b.String()
```

**Review by louisruch — APPROVED** [review]
one comment re the capabilities stringer but otherwise this looks great

File: internal/credential/vault/vault_capabilities.go (hunk 4 above)
Comment (mgaffney) [reply internal/credential/vault/vault_capabilities.go:114]: Great idea. Ty!

**Review by louisruch — APPROVED** [review]
LGTM

**Pull request merged**